로또 2단계 미션 제출합니다.

chanhalee opened the pull request:
안녕하세요. seokhong님!
지난번 리뷰 정말 감사했습니다!
석홍님의 상세한 리뷰 덕분에 많은 도움을 받았습니다!
이번에도 잘 부탁드립니다!

parkjoos commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -12,11 +14,24 @@
+        this.lottoNumbers = Collections.unmodifiableSet(lottoNumbers);
+    }
+
+    public static Lotto autoGenerate(LottoNumberGenerator lottoNumberGenerator) {
+        Set<LottoNumber> newLottoNumbers = getEmptyLottoNumbers();
+        newLottoNumbers.addAll(lottoNumberGenerator.getNumbers());
+        return new Lotto(newLottoNumbers);
+    }
+
+    public static Lotto manualGenerate(String numbersText) {
+        Set<LottoNumber> newLottoNumbers = getEmptyLottoNumbers();
+        newLottoNumbers.addAll(LottoNumberParser.toLottoNumbers(numbersText));
+        return new Lotto(newLottoNumbers);
+    }
+
+    private static Set<LottoNumber> getEmptyLottoNumbers() {
+        return new TreeSet<>(Comparator.comparingInt(LottoNumber::getNumber));
     }
```
이 세가지 메소드가 꼭 이 Lotto 도메인에 있어야 할 필요가 있을까요?
도메인 클래스내의 멤버함수는 그 객체를 조작하거나 객체 내부 데이터를 읽을 필요가 있을때에만 사용하도록 하고
이런 부분은 controller 나 util 성 클래스로 분리하는게 어떨까요?
아니면 이 함수들이 꼭 필요한 부분인지 다른 부분으로 합쳐서 더 간소화 할 방법은 없는지도 생각해 보면 좋을것 같습니다.

parkjoos commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -27,10 +25,35 @@ public LottoController() {
+    private List<Lotto> buyAutoLotto(NumOfLotto numOfAuto, LottoNumberGenerator lottoNumberGenerator) {
+        List<Lotto> lottoList = new ArrayList<>();
+        for (int i = 0; i < numOfAuto.getNumOfLotto(); i++) {
+            lottoList.add(Lotto.autoGenerate(lottoNumberGenerator));
```
```suggestion
            lottoList.add(new Lotto(lottoNumberGenerator.generate()));
```
또는 
```suggestion
            lottoList.add(Lotto.of(lottoNumberGenerator.generate()));
```

정도로만 하면 Lotto 내부의 로직도 정리되고 좀더 간결해지지 않을까요?

32sylee replied:
생각해보니 불필요한 메서드 같네요..! ManualNumberGenerator 관련 답변주신 내용이랑 같이 고민해서 개선해보겠습니다:) 감사합니다!!

parkjoos commented:
질문에서 말씀 하신 것처럼 LottoNumberGenerator 를  구현하는 새로운 ManualNumberGenerator 를 만들어 볼 수도 있을것 같습니다.
ManualNumberGenerator 에 새로운 멤버변수로 inputString 을 받을 수 있을 것 같구요
Parser 의 사용을 ManualNumberGenerator 안으로 숨길수는 있을 것 같습니다.
그리고 generate 될때 그냥 사용자에게 입력받은 문자를 Set<LottoNumber> 로 변환해서 리턴 해주는 역할 정도일것 같네요

parkjoos reviewed (changes requested):
안녕하세요 jay 입니다. 😁
1단계에서 구조를 잘 잡아 두었기 때문에 2단계 기능을 추가하는데에 크게 어려움이 있지는 않았을 것 같아요
이번 단계에서 시에라가 한가지 연습해 봤으면 하는게 있는데요
새로운 기능에 대해 코드를 추가/수정 하면서 기존 로직에 영향을 끼치지 않는지 알아보는 가장 좋은 방법은
기존에 작성된 테스트 코드를 전부다 돌려 보는 것 입니다.
2단계 PR 을 받아서 전체 테스트 코드를 돌려보니 실패하는 테스트가 몇개 있네요 
항상 코드를 작성하거나 새 테스트 케이스를 만들면 그 테스트 코드만 실행 해 보는것이 아니라
프로젝트의 전체 테스트 코드를 주기적으로 확인 해 보는것이 좋습니다. 
테스트 케이스는 만들어 놓기만 하면 되는게 아니라 항상 전부 통과가 되는지 확인이 되는것도 매우 중요합니다.
코멘트들 한번 확인 해보시고 
지금 실패하는 테스트 케이스들도 한번 확인해주세요

32sylee commented on src/main/java/lotto/util/LottoNumberParser.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.stream.Collectors;
+
+public class LottoNumberParser {
+    public static final String SPLIT_DELIMITER = ", ";
```
trim을 하는 방법이 있었군요!! 수정하겠습니다:)

32sylee commented on src/test/java/lotto/domain/MatchNumberTest.java:
```diff
@@ -1,22 +1,33 @@
+        assertThatNoException().isThrownBy(() -> {
+            MatchNumber.of("1, 2, 3, 4, 5, 6", 6);
+        });
+    }
```
헉 예외가 발생하는 걸 검증해야 하는데 반대로 적었네요...ㅠㅠㅠ 수정했습니다!

32sylee commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -17,8 +17,8 @@ public int getMoney() {
 
-    public int howMany() {
-        return money / LOTTO_PRICE;
+    public NumOfLotto howMany() {
```
PR을 하기 전에 테스트를 돌렸어야 했는데 그냥 PR을 해버렸군요ㅠㅠ
앞으로는 잊지 말고 전체 테스트 다 돌려보겠습니다! 해당 부분 수정했습니다:)

32sylee commented on src/main/java/lotto/domain/WinningNumber.java:
```diff
@@ -16,25 +14,8 @@ private WinningNumber(Set<LottoNumber> winningNumber){
+                LottoNumberParser.toLottoNumbers(numbersText));
     }
 
     public Set<LottoNumber> getWinningNumber(){
```
넵 메서드 삭제했습니다!

32sylee commented on src/main/java/lotto/domain/NumOfLotto.java:
```diff
@@ -0,0 +1,24 @@
+
+    public NumOfLotto minus(NumOfLotto other) {
+        return new NumOfLotto(getNumOfLotto() - other.getNumOfLotto());
+    }
```
넵! 테스트 추가하겠습니다

32sylee commented:
jay 안녕하세요! 피드백 감사합니다😊
코멘트 주신대로 Lotto 클래스에서 generate하는 부분 제거하고 Ganerator 클래스로 통합했습니다.
그리고 기존에 LottoNumberGenerator 인터페이스를 구현받아 사용하던 부분은 지금 코드에서는 크게 필요가 없는 것 같아서
Generator 클래스들은 util성 클래스로 변경했습니다!

parkjoos commented on src/main/java/lotto/util/LottoNumberParser.java:
```diff
@@ -3,15 +3,16 @@
-        Set<LottoNumber> lottoNumbers = Arrays.stream(split(numbersText))
+        Set<LottoNumber> lottoNumbers = trim(split(numbersText)).stream()
                 .map(number -> LottoNumber.of(getParseInt(number)))
                 .collect(Collectors.toSet());
```
trim 에서 map 과 collect 를 다 한다음에 다시 map 과 collect 를 하고 있는데
이건 따로따로 순회를 두번한 거라 stream 을 사용한 이점이 없어진것 같아요
map 안에서 한번에 처리 할 수 있을것 같은데 다음 미션에서 이런 상황이 나오면 하나의 stream 에서 되도록 연습해 보세요

32sylee replied:
하나의 stream에서 trim하고 split을 함께 처리할 수 있겠네요..!!
다음번 미션 할 때 기억해서 연습해보겠습니다. 코멘트 감사합니다:D

parkjoos reviewed (approved):
리뷰 반영 아주 잘 해 주셨습니다. 
간단하게 코멘트 드린건 있는데요 이번 미션은 머지 해도 될 것 같아요
다른 미션에서 코멘트 내용에 대해 연습해 보면 좋을 것 같습니다. 
그럼 로또 미션 정말 수고 하셨습니다. 👏 다음 미션도 화이팅 입니다.

chanhalee committed:
docs(README) Todo List 작성

Todo List 작성

chanhalee committed:
feat(Budget) 구매 금액에 대한 요구사항 Test 작성 및 구현

구매 금액에 대한 요구사항 Test 작성 및 구현

chanhalee committed:
feat(Ball) 로또 번호에 대한 요구사항 test 작성 및 구현

로또 번호에 대한 요구사항 test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 정수여야 한다.

chanhalee committed:
feat(Ticket) 티켓에 대한 요구사항 Test 작성 및 구현

티켓에 대한 요구사항 Test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 중복되지 않는 6개의 수로 이루어져 있다

chanhalee committed:
refactor(Budget) validation 메서드 static 속성 제거

validation 메서드 static 속성 제거

chanhalee committed:
feat(WinningNumbers) 티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현

티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현
- 등수에 대한 Enum 클래스 Prize 작성
- Ball -> Number로 클래스 명 변경
- Number Set을 보유하는 Numbers 클래스 생성
- WinningNumbers, TicketNumbers가 Numbers를 상속 받도록 구현

chanhalee committed:
feat(PrizeCalculator) 수익률계산기 테스트 작성 및 구현

수익률계산기 테스트 작성 및 구현

chanhalee committed:
feat(LottoMachine) LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

chanhalee committed:
feat(LottoMachine) LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

chanhalee committed:
refactor MVC 패턴에 맞게 패키지 분리

chanhalee committed:
feat(LottoView) view 레이어 구현

view 레이어 구현

chanhalee committed:
feat(LottoController) Controller 구현

Controller 구현

chanhalee committed:
Refactor

리뷰어의 코멘트를 반영하여 리펙터링

chanhalee committed:
feat(ManuallyChosenNumbers) 로또 수동을 위한 도메인 클래스 구현

로또 수동을 위한 도메인 클래스 구현

chanhalee committed:
feat(LottoController) 로또 수동 입력 기능을 컨트롤러에 구현

로또 수동 입력 기능을 컨트롤러에 구현
- 네이밍 일관성을 위해 TicketNumbers -> AutoGeneratedTicketNumbers로 클래스명 변경
- 수동 입력할 로또의 개수를 차감하기 위해 Budget 내부 수정
- 수동 입력된 로또를 LottoMachine에 등록하기 위한 메서드 구현
- 수동 입력과 에러에 대한 View 메서드 추가

chanhalee committed:
feat(LottoController) 로또 수동 입력 기능을 컨트롤러에 구현

로또 수동 입력 기능을 컨트롤러에 구현
- 네이밍 일관성을 위해 TicketNumbers -> AutoGeneratedTicketNumbers로 클래스명 변경
- 수동 입력할 로또의 개수를 차감하기 위해 Budget 내부 수정
- 수동 입력된 로또를 LottoMachine에 등록하기 위한 메서드 구현
- 수동 입력과 에러에 대한 View 메서드 추가

chanhalee committed:
Merge branch 'master' into step2

# Conflicts:
#	README.md
#	src/main/java/lotto/controller/LottoController.java
#	src/main/java/lotto/domain/Budget.java
#	src/main/java/lotto/domain/LottoMachine.java
#	src/main/java/lotto/domain/WinningNumbers.java
#	src/main/java/lotto/view/LottoView.java
#	src/test/java/lotto/domain/LottoMachineTest.java
#	src/test/java/lotto/domain/WinningNumbersTest.java

chanhalee committed:
refactor

Number#numberInstances 를 한번에 초기화하는 메서드 구현

chanhalee committed:
refactor

WinningNumbers 가 Numbers를 상속받지 않고 조합으로 인스턴스 변수로 보유하는 방향으로 리팩터링

seokhong reviewed (changes requested):
찬하님, 2단계 미션 잘 구현하셨습니다. 👍 
고민해보시면 좋을 것 같은 내용 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙂

seokhong reviewed (approved):
찬하님, 로또 미션 진행하는 동안 수고 많으셨습니다.
피드백 작성을 위해 PR은 머지하겠습니다.
즐거운 한 주 보내세요. 🙂
